README.md
# Stepruder

This is a pen-testing tool that implements Intruder-like functionality in Python. In addition, it sends a sequence of requests for every payload iteration. The result is a combination of Stepper (new Burp extenstion) and Intruder functionlity.

## Getting Started

Complete the following instructions to set up and run your project on your local machine for development and testing purposes. No special deployment for production environment is provisioned.

Getting the source:
```
git clone https://github.com/blackberry/Stepruder
```

### Installing

Python3 installation is required as this project is a python script designed to run under Python3. Tested on Ubuntu / CentOS / Windows.

### Usage 

Usage: stepruder.py [-h] [-s SEPARATOR] [-l LOG] [-v] requestsfile configfile

This script automates payload injection into the long request sequences rather than into an individual request. It parses *requestsfile* and *configfile* json. Then it applies substitutions from the config file, injects the payloads and sends the processed requests in order.

Positional arguments:
* requestsfile - Text file containing sequence of HTPP requests, each separated by separator line (default #####)
* configfile - JSON file containing variables substitutions and other configs

Optional arguments:
* -h, --help            show this help message and exit
*   -s SEPARATOR, --separator SEPARATOR Custom separator between requests in requestsfile
*   -l LOG, --log LOG     Traffic log for debug purposes
*   -v, --verbose         Increase output verbosity

```
python stepruder.py -l debug.log -v requests/requests_login.txt configs/config_login.json
```

### How this works

Requestsfile should contain a plaintext sequence of request templates with potential substitution variables and eval expressions. 
Substitution variables can be three different scopes:
1. Constant substitutions
2. Sequence-scope substitutions
3. Request-scope substitutions

**Constant substitutions** are defined under "substitutions" in config json as "constants" and replaced only once before the start of the sending process. This will be a simple string replacement substitution. These are useful for variables that stay constant through the entire program but may be 
different between the program runs (i.e. cookie).

**Sequence-scope substitutions** are defined in config json as "payloads" and replaces before the beginning of each sequence. The number of different payloads will define the number of sequence iterations. Lack of any payloads will only send the sequence once. These are useful to define injected payloads.

**Request-scope substitutions** are defined under "substitutions" in config json as "responsevars", because they are defined as matches from response content and therefore can change after every response. The match can be "json" or "regex". Use "json" when working only on json response bodies and use "regex" when working on any response body plus response header values, in that order. These substitutions are potentially replaced before sending every new request depending on whether the previous response contained the match or not. If the response does not contain the match, then the responsevar value does not change. Responsevars are useful to keep interaction between client and server alive (i.e. updating CSRF tokens, keeping client-server sync in Vaadin etc.).

**Eval expression** are substitutions on steroids. Eval expressions enclosed in ${#...} are sequence-scoped (i.e. evaluated once in the beginning of the sequence). eval expressions enclosed in ${!...} are request-scoped. Eval expression by itself is a Python eval expression that returns 
one value. This value is then substituted into the request sequence. The eval expression may or may not include variables (i.e. ```${!time.time()}``` vs ```${#len(PAYLOAD)}```). The expression should only use Python modules imported already in the script (see list at the beginning of the stepruder.py) or import its own modules.

The simplified overall flow of the program is as follos:

```
Parsing requestfile and config json
        Form payloads arrays, identify number of sequences
        Constant substitutions
        While more sequences - start sequence i:
                Payload #1 injection
                Evaluation of sequence-scope eval expressions
                While more requests in the sequence:
                        Substitute responsevars from j-1 response
                        Evaluate request-scope eval expressions
                        Send request j
                        Parse responsevars from j response
```

In addition, config json may include the following connection configs: ssl (Boolean), port (Numeric), proxies (Dictionary) and response parsing regex "grep_final_response". For example:
```
        "ssl": true,
        "port": 443,
        "grep_last_response": "Invalid credentials",
        "proxies": {
                "http": "https://127.0.0.1:8080"
                "https": "https://127.0.0.1:8080"
        },
        "substitutions": {...
```

### Limitations

Current Stepruder limitations:
1. No parallel sequence sending yet. This should improve the performance dramatically.
2. No different payload injection combos (i.e. only Sniper or Pitchfork in Intruder terms).
3. No advanced encoding/decoding capabilities yet. Payloads / substitutions that include special characters might interfere with JSON parsing and regex matching.

## Contributing

Please read [CONTRIBUTING.md](https://gist.github.com/PurpleBooth/b24679402957c63ec426) for details on our code of conduct, and the process for submitting pull requests to us.

## Authors

* **Shay Berkovich** - *Initial work* - [sshayb](https://github.com/sshayb)

## License

This project is licensed under the Apache 2.0 License - see the [LICENSE](LICENSE) file for details.



CHANGELOG
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.0.1] - 2019-11-14
### Added

- Implementing request file parsing
- Implementing config file parsing
- Basic sending / substitution logic
- Initial commit

## [0.0.2] - 2020-01-23
### Added

- Fixing bugs
- Fixing printing errors
- Making regex response parsing work
- utf8 encoding of POST data

## [0.0.3] - 2020-01-23
### Added

- Re-establishing the process flow - sequence-scope vs request-scope substitutions
- ${#} vs ${!}
- Adding eval functionality and refactoring the substitutions logic - using regex sub function heavily
- Cleaning up the exception/program termination logic - when to terminate when to continue

## [0.0.4] - 2020-01-30
### Added

- Verbose help output
- Updating README
- Adding proxy option for request sending
- Adding grep_last_Response feature
- Refactoring globals.py

## [0.0.5] - 2020-04-06
### Added

- Adding encoding functionality for payloads with optionsl url- and jsonencoding
- Cleaning up some error and exception code
- Increasing requests timeout to 4 sequence
- Fixing proxies bug in GET request
- Higher logging level in none-verbose mode

globals.py
################################################################################
# Name   : Stepruder - globals definitions
# Author : Shay Berkovich
#
# Copyright 2020 BlackBerry Limited
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.
################################################################################

#GLOBAL VARS
constant_substitutions = {}
payloads = {}
separator = None
grep_last_response_regex = None
sslconfig = False
port = 80
proxies = {}

help_string = "Requestsfile should contain a plaintext sequence of request templates with potential substitution \
variables and eval expressions. Substitution variables can be of 3 different scopes: \n(1) Constant substitutions \
\n(2) Sequence-scope substitutions \n(3) Request-scope substitutions \n\n\
Constant substitutions defined under \
\"substitutions\" in config json as \"constants\", are replaced only once before the start of the sending process. \
This will be a simple string replace substitution. These are useful for variables that stay constants through \
entire program but may be different between the program runs (i.e. cookie). \n\n\
Sequence-scope substitutions defined \
in config json as \"payloads\", are replaces before the beginning of each sequence. Number of different payloads \
will define the number of sequence iterations. Lack of any payloads will only send the sequence once. These are \
useful to define injected payloads.\n\n\
Request-scope substitutions defined under \"substitutions\" in config json as \"responsevars\", because the are defined\
 as matches from response content and as such can change after every response. The match can be \"json\" or \"regex\" \
with json working only on json response bodies and regex working on any response body plus response header values, in \
that order. These substitutions are potentially replaced before sending every new request depending on whether the previous \
response contained the match or not. If the response does not contain the match, then the responsevar value does not \
change. Responsevars are useful to keep alive interaction between client and server (i.e. updating CSRF tokens, \
keeping client-server sync in Vaadin etc.)\n\n\
Eval expression are substitutions on steroids. Eval expressions enclosed in ${#...} are sequence-scoped\
 (i.e. evaluated once in the beginning of the sequence); eval expressions enclosed in ${!...} are request-scoped. Eval\
 expression by itself is a Python eval expression that returns one value. This value is then substituted into the \
request sequence. The eval expression may or may not include variables (i.e. ${!time.time()} vs ${#len(PAYLOAD)}). \
The expression should only use Python modules imported already in the script (see list at the \
beginning of the stepruder.py) or import its own modules.\n\n\
The simplified overall flow of the program looks as following:\n\
\tParsing requestfile and config json\n\
\tForm payloads arrays, identify number of sequences\n\
\tConstant substitutions\n\
\tWhile more sequences - start sequence i:\n\
\t\tPayload #1 injection\n\
\t\tEvaluation of sequence-scope eval expressions\n\
\t\tWhile more requests in the sequence:\n\
\t\t\tSubstitute responsevars from j-1 response\n\
\t\t\tEvaluate request-scope eval expressions\n\
\t\t\tSend request j\n\
\t\t\tParse responsevars from j response\n\
\
\n\nIn addition, config json may include following connection configs: ssl (Boolean), port (Numeric), proxies \
(Dictionary) and response parsing regex \"grep_final_response\". For example:\n\
\t\"ssl\": true,\n\
\t\"port\": 443,\n\
\t\"request_separator\": \"#####\",\n\
\t\"grep_last_response\": \"Invalid credentials\",\n\
\t\"proxies\": {\n\
\t\t\"http\": \"https://127.0.0.1:8080\"\n\
\t\t\"https\": \"https://127.0.0.1:8080\"\n\
\t},\n\
\t\"substitutions\": {...\
\
\n\nCurrent Stepruder limitations:\n\
(1) No parallel sequence sending yet. This should improve the performance dramatically.\n\
(2) No different payload injection combos (i.e. only Sniper or Pitchfork in Intruder terms).\n\
(3) No advanced encoding/decoding capabilities yet. Payloads / substitutions that include special characters \
might interfere with JSON parsing and regex matching.\n"

sequencer.py
#!/usr/bin/python
import sys
import argparse
import re
import json
import requests
import urllib3

from http.server import BaseHTTPRequestHandler
from io import BytesIO

#cass for parsing the HTTP request from text - adapted from https://stackoverflow.com/questions/4685217/parse-raw-http-headers
class HTTPRequest(BaseHTTPRequestHandler):
    def __init__(self, request_text):
        self.rfile = BytesIO(request_text)
        self.raw_requestline = self.rfile.readline()
        self.error_code = self.error_message = None
        self.parse_request()

    def send_error(self, code, message):
        self.error_code = code
        self.error_message = message

urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)


#parse payloads from config
#returns number of iterations based on minimal number of payload lists
def parse_payloads():
	#perform substitutions and do a substitution requestbulk
	if config['payloads']:
		iterations = sys.maxsize
		for key in config['payloads']:
			payloads[key] = []
			if config['payloads'][key]['type'] == 'numeric':
				#parse payloads from the given range
				for i in range(int(config['payloads'][key]['start']), int(config['payloads'][key]['end']), int(config['payloads'][key]['step'])):
					payloads[key].append(i)
			elif config['payloads'][key]['type'] == 'list':
				#parse payloads from the list
				try:
					with open(config['payloads'][key]['path'], 'r', encoding='utf-8') as payloadlist:
						line = payloadlist.readline()
						while line and line != "":
							#stripping from newlines and whitespaces before adding
							payloads[key].append(line.strip())
							line = payloadlist.readline()
				except:
					print ("Can't parse/understand payloads in config json - can't open list", config['payloads'][key1]['path'])
					sys.exit(0)
			else:
				print ("Can't parse/understand payloads in config json - unknown payload type")
				sys.exit(0)
			iterations = iterations if iterations < len(payloads[key]) else len(payloads[key])
	else:
		iterations = 0
	return iterations
		
#send signle request given the string request
def send_request(rstr, sequence_number, iteration_number):
	
	#embed payloads
	for pkey, pvalue in payloads.items():
		#print (pkey, pvalue[iteration_number])
		rstr = rstr.replace(pkey, str(pvalue[iteration_number]))
	
	#automatic request parsing
	rbytes = rstr.strip().encode('utf-8')
	try:
		request = HTTPRequest(rbytes)
	except:
		print ("Can't parse/understand HTTP requests from file")
		return
	
	#forming the url line - assuming first word is always method
	if request.requestline.split(" ")[1].startswith('http'):
		#full url is already included in the requestline
		url = request.requestline.split(" ")[1]
	else:
		#requestline is only a path + queryargs
		if sslconfig:
			url = "https://" + request.headers['host'].split(":")[0] + ":" + str(port) + request.requestline.split(" ")[1]
		else:
			url = "https://" + request.headers['host'].split(":")[0] + ":" + str(port) + request.requestline.split(" ")[1]

	if (args.verbose):
		print ('Sending {0} request {1} to {2}'.format(request.command, sequence_number, url))

	try:
		if (request.command == 'GET'):
			r = requests.get(
				url,
				headers=request.headers,
				verify=False)
		elif (request.command == 'POST'):
			#get data first, if PUT/PUSH methods are added move this out of the branching statement
			#according to RFC, we expect at least one line in the request body 
			rdata = rstr.split("\n\n")[1]
				
			r = requests.post(
				url,
				headers=request.headers, 
				data=rdata,
				verify=False)
		else:
			#other methods are easy to add
			print ("Unsupported method")
			return
	except:
			print ("Error sending request " + str(sequence_number))
			return

	#print (r.request.body)
	if (args.verbose):
		print ("Status:{0}#####Content length:{1}#####Response time:{2}#####Content:{3}".format(r.status_code, 
				len(r.content), 
				r.elapsed.total_seconds(),
				r.content))
	else:
		print ("{0}#####{1}".format(r.status_code, len(r.content)))
		
		
#send the sequence of requests given the array of string requests
def send_sequence(request_list, iteration):
	num = 0
	print ("Iteration " + str(iteration) + " - starting sequence")
	for rstr in request_list:
		num+=1
		send_request(rstr, num, iteration)
	print ("Iteration " + str(iteration) + " - sequence finished")
	
	
#MAIN FUNCTION	
# Argument parsing
parser = argparse.ArgumentParser(description='Script that parses requests file and configuration json. It applies substitutions from config file and sends the requests in order.')
parser.add_argument("requestsfile", help="Text file containing sequence of HTPP requests, each separated by separator line (default #####)")
parser.add_argument("configfile", help="JSON file containing variables substitutions and other config")
parser.add_argument("-s", "--separator", help="custom separator between requests in requestsfile")
parser.add_argument("-v", "--verbose", action="store_true", help="increase output verbosity")
args = parser.parse_args()

try:
	with open(args.requestsfile, 'r') as requestsfile:
   			requestbulk = requestsfile.read()
except:
	print ("Can't open requests file", args.requests)
	sys.exit(0)

#parse config file as json
#try:
with open(args.configfile, 'r') as configfile:
	config = json.load(configfile)
		#config = configfile.read()
#except:
#	print ("Can't load json from file", args.configfile)
#	sys.exit(0)

#parse the ssl config
sslconfig = False
port = 80
if 'ssl' in config and config['ssl']:
	sslconfig = True
	port = 443
if 'port' in config:
	port = config['port']

#substitutions are perofmed once before starting the iterations
substitutions = config['substitutions']
for key, value in substitutions.items(): 
	#we can assume the format for substitution is "key: value"
	requestbulk = requestbulk.replace(key, value)

#parse the bulk and get a list of requests
if args.separator:
	request_list = re.split(args.separator, requestbulk)
else:
	request_list = re.split(r'#####', requestbulk)
	
payloads = {}
if 'payloads' in config.keys():
	iterations = parse_payloads()
else:
	iterations = 1

if (args.verbose):
	print ("Parsed: {0} substitutions and {1} injection points; payload sequence length is {2}; total requests to be sent {3}.".format(
		len(config['substitutions']), 
		len(payloads), 
		iterations,
		len(request_list)*iterations))

for i in range(0, iterations, 1):
	send_sequence(request_list, i)
	

stepruder.py
################################################################################
# Name   : Stepruder - main
# Author : Shay Berkovich
#
# Copyright 2020 BlackBerry Limited
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.
################################################################################

#!/usr/bin/python
import sys
import argparse
import re
import json
import requests
import urllib3
import urllib.parse
import logging
import time
import http.client

from collections import defaultdict
from http.server import BaseHTTPRequestHandler
from io import BytesIO
from globals import *

#class for parsing the HTTP request from text - adapted from https://stackoverflow.com/questions/4685217/parse-raw-http-headers
class ParsedHTTPRequest(BaseHTTPRequestHandler):
    def __init__(self, request_text):
        self.rfile = BytesIO(request_text)
        self.raw_requestline = self.rfile.readline()
        self.error_code = self.error_message = None
        self.parse_request()

    def send_error(self, code, message):
        self.error_code = code
        self.error_message = message

#helper function to evaluate expression
def repl(m):
    expr = m.group(1)
    return str(eval(expr))

#parse payloads from config
#returns number of iterations based on minimal size of payload lists
def parse_payloads():
	if config['payloads']:
		iterations = sys.maxsize
		for key in config['payloads']:
			payloads[key] = []
			if config['payloads'][key]['type'] == 'numeric':
				#parse payloads from the given range
				for i in range(int(config['payloads'][key]['start']), int(config['payloads'][key]['end']), int(config['payloads'][key]['step'])):
					payloads[key].append(i)
			elif config['payloads'][key]['type'] == 'list':
				#prep encoding if needed
				if config['payloads'][key]['encoding']:
					encodingsequence = config['payloads'][key]['encoding'].split(',');
				#parse payloads from the list
				try:
					with open(config['payloads'][key]['path'], 'r', encoding='utf-8') as payloadlist:
						line = payloadlist.readline()
						while line and line != "":
							#payload processing befire sending
							lineprocessed = line.strip()
							for enc in encodingsequence:
								if enc == "urlencode":
									lineprocessed = urllib.parse.quote(lineprocessed)
								if enc == "jsonencode":
									lineprocessed = json.dumps(lineprocessed)
							payloads[key].append(lineprocessed)
							line = payloadlist.readline()
				except:
					print ("Error - can't parse/understand payloads in config json - can't open list {0}, exiting.".format(config['payloads'][key]['path']))
					sys.exit(1)
			else:
				print ("Error - can't parse/understand payloads in config json - unknown payload type, exiting.")
				sys.exit(1)
			iterations = iterations if iterations < len(payloads[key]) else len(payloads[key])
	else:
		iterations = 0
	return iterations

#parse the response and retrieve the value based on json key
#if multiple values appear the behavior is undefined
def retrieve_responsevar_json(resp, json_exp):
	try:
		respdict = json.loads(resp.text)	#json parse just the body of the response
		value = respdict[json_exp]	#TODO: what happens if 2 duplicate keys present? Python dictionary does not support duplicate keys BTW
		return value
	except Exception as e:
		#return none if something goes south
		if args.verbose:
			print ("Error trying to match JSON response: {0} with expression {1}, exiting.".format(resp, json_exp))
		sys.exit(1)

#parse the response and retrieve the value based on the first regex match
#if capturing group is defined in the regex - the first matching capturing group is captured
#if capturing group is not defined in the regex - the whole match is returned
def retrieve_responsevar_regex(resp, regex_exp):
	try:
		#first search the body
		value = re.search(regex_exp, resp.text)
		if value:
			if args.verbose:
				print ('Match for {0} : {1}'.format(regex_exp, value.group(1)))
			return value.group(1)
		#if nothing matches - search the headers
		for headername, headerval in resp.headers.items():
			value = re.search(regex_exp, headerval)
			if value:
				if args.verbose:
					print ('Match for {0} : {1}'.format(regex_exp, value.group(1)))
				return value.group(1)
		if args.verbose:
			print ('No match for {0}'.format(regex_exp))
	except Exception as e:
		#return none if something goes south
		if args.verbose:
			print ("Error trying to match regex {0} in response {1}, exiting.".format(regex_exp, resp.text))
		sys.exit(1)

#send signle request given the request string
def send_request(rstr, request_number, iteration_number):
	
	#embed responsevar substitutions if any
	#in first request first iteration there should be no substitutions
	for skey, svalue in current_step_substitutions.items():
		if svalue:
			if args.verbose:
				print ("Considering substitutions: " + str(skey) + " with " + str(svalue))
			rstr = rstr.replace(skey, str(svalue))
			if args.verbose:
				print ('Replacement in current iteration request: replaced {0} with {1}'.format(skey, svalue))
	
	#evaluate request-scope expressions - starting with ${!
	try:
		rstr = re.sub("\\${!(.*?)}", repl, rstr)
	except Exception as e:
		print ("Error evaluating request-scope expression in sequence {0}, request{1}, exiting.".format(str(iteration_number), request_number))
		sys.exit(1)
									
	#automatic request parsing
	rbytes = rstr.strip().encode('utf-8')
	try:
		request = ParsedHTTPRequest(rbytes)
	except Exception as e:
		print ("Error trying parse/understand HTTP requests from file, exiting: ", e)
		sys.exit(1)
	
	#parser can fail without throwing an exception
	if request.error_message:
		print ("Error trying parse/understand HTTP requests from file: ", request.error_message)
		sys.exit(1)

	#forming the url line - assuming first word is always method
	if request.requestline.split(" ")[1].startswith('http'):
		#full url is already included in the requestline
		url = request.requestline.split(" ")[1]
	else:
		#requestline is only a path + queryargs
		if sslconfig:
			url = "https://" + request.headers['host'].split(":")[0] + ":" + str(port) + request.requestline.split(" ")[1]
		else:
			url = "http//" + request.headers['host'].split(":")[0] + ":" + str(port) + request.requestline.split(" ")[1]

	if (args.verbose):
		print ('Sending {0} request {1} to {2}'.format(request.command, request_number, url))

	try:
		if logfile:
			logfile.write(request.command + " ")
			logfile.write(url + "\n")
			logfile.write(str(request.headers))

		if (request.command == 'GET'):
			r = requests.get(
				url,
				headers=request.headers,
				verify=False,
				proxies=proxies,
				timeout=(4,4))	#2s should to connect and read the response TODO should this be in config?
		elif (request.command == 'POST'):
			#get data first, if PUT/PUSH methods are added move this out of the branching statement
			#according to RFC, we expect at least one line in the request body 
			#limiting number of spits to 1 to only split on the first newline in case there are multiple
			rdata = rstr.split("\n\n", 1)[1].strip().encode('utf-8')
			if logfile:
				logfile.write(str(rdata) + "\n")

			r = requests.post(
				url,
				headers=request.headers, 
				data=rdata,
				verify=False,
				proxies=proxies,
				timeout=(4,4)) #TODO should timeout be part of the config?
		else:
			#other methods are easy to add
			print ("Unsupported method")
			current_step_substitutions.clear()	#no response - no substitutions
			return
	except Exception as e:
			print ("Error sending request {0}, exiting. Exception: {1}".format(request_number,e))
			sys.exit(1)

	#fill up current (next) step substitutions
	#TODO: if this is the last response in the sequence - carry over to the next iteration?
	for skey, svalue in step_substitutions[(request_number-1) % len(request_list)].items():
		if svalue[0] == 'json':
			#for json it only makes sense to scan the response body
			new_value = retrieve_responsevar_json(r, svalue[1])
			#we are only interested in non-trivial values, otherwise leave it the same substitution
			if new_value != None:
				current_step_substitutions[skey] = new_value
		elif svalue[0] == 'regex':
			#for regex we want to scan both the body and headers (f.e. cookies)
			new_value = retrieve_responsevar_regex(r, svalue[1])
			#we are only interested in non-trivial values, otherwise leave it the same substitution
			if new_value != None:
				current_step_substitutions[skey] = new_value
		else:
			print ("Error while performing substitutions - unrecognized response var, continuing.")
	
	print ("{0}#####{1}".format(r.status_code, len(r.content)))
	if logfile:
		logfile.write("---------------------\n" + str(r.content) + "\n======================\n")

	# if last payload and grep_result is present check for response
	if grep_last_response_regex and request_number == len(request_list):
		value = re.search(grep_last_response_regex, r.text)
		if value: 
			print ("Grep match found in last response: " + value.string)
		else:
			print ("Grep match not found in last response.")

#send the sequence of requests given the array of string requests
def send_sequence(request_list, iteration):
	num = 0
	print ("Iteration " + str(iteration) + " - starting sequence")
	
	#first request does not have response - init the current_step_substitutions with trivial value
	current_step_substitutions.clear()
	
	request_list_post_sequence_scope__subs = []
	for rstr in request_list:	
		try:
			for pkey, pvalue in payloads.items():
				#we are doing json escaping when needed at the payload processing stage
				#s = str(pvalue[iteration]).replace('"', '\\"').replace('\\', '\\\\')
				try:
					rstr = re.sub(pkey, str(pvalue[iteration]), rstr)
				except Exception as e:
					print ("Can't inject payload - bad character? Continuing.")
					current_step_substitutions.clear()	#no response - no substitutions
					return
			rstr = re.sub("\\${#(.*?)}", repl, rstr)
		except Exception as e:
			print ("Error matching sequence-scope regex in sequence {0}, exiting: {1}".format(iteration, e))
			sys.exit(1)	
		
		request_list_post_sequence_scope__subs.append(rstr)

	print ("Status#####ResponseLen")	
	for rstr in request_list_post_sequence_scope__subs:
		num+=1
		send_request(rstr, num, iteration)

	print ("Iteration " + str(iteration) + " - sequence finished")
	if logfile:
		logfile.write("\n\n********************************************** End of the iteration **********************************************\n\n")
	
#MAIN FUNCTION	
# Argument parsing
parser = argparse.ArgumentParser(
	description='This is a mix between Stepper and Intruder script that automates payload injection into the long request sequences rather than into an individual request. It parses requests file and configuration json. It applies then substitutions from the config file, injects the payloads and sends the processed requests in order.',
	formatter_class=argparse.RawTextHelpFormatter,
	epilog=help_string)
parser.add_argument("requestsfile", help="Text file containing sequence of HTPP requests, each separated by separator line (default #####)")
parser.add_argument("configfile", help="JSON file containing variables substitutions and other configs")
parser.add_argument("-l", "--log", help="Traffic log for debug purposes")
parser.add_argument("-v", "--verbose", action="store_true", help="Increase output verbosity")
args = parser.parse_args()

#networking and logging config
urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
if args.verbose:
	http.client.HTTPConnection.debuglevel = 1
	logging.basicConfig()
	logging.getLogger().setLevel(logging.DEBUG)
	requests_log = logging.getLogger("requests.packages.urllib3")
	requests_log.setLevel(logging.DEBUG)
	requests_log.propagate = True
else:
	http.client.HTTPConnection.debuglevel = 0
	logging.basicConfig()
	logging.getLogger().setLevel(logging.CRITICAL)
	requests_log = logging.getLogger("requests.packages.urllib3")
	requests_log.setLevel(logging.CRITICAL)
	requests_log.propagate = True

#read requests file
try:
    with open(args.requestsfile, 'r') as rf:
        requestbulk = rf.read()
except Exception as e:
	print ("Error - can't open requests file {0}, exiting.".format(args.requestsfile))
	sys.exit(1)

#parse config file as json
try:
    with open(args.configfile, 'r') as configfile:
        config = json.load(configfile)
		#config = configfile.read()
except Exception as e:
	print ("Error - can't load json from file {0}, exiting.".format(args.configfile))
	sys.exit(1)

try:
	logfile = None
	if args.log:
		logfile = open(args.log, 'w')
except Exception as e:
	print ("Error - can't open log file {0} for writing, exiting.".format(args.log))
	sys.exit(1)

#parse the ssl config
if 'ssl' in config and config['ssl']:
	sslconfig = True
	port = 443
if 'port' in config:
	port = config['port']
if 'request_separator' in config:
	separator = config['request_separator']
if 'proxies' in config and config['proxies']:
	proxies = config['proxies']

#constant substitutions are performed once before starting the iterations
if 'substitutions' in config.keys() and 'constants' in config['substitutions'].keys():
    constant_substitutions = config['substitutions']['constants']
for key, value in constant_substitutions.items(): 
	#we can assume the format for substitution is "key: value"
	if args.verbose:
		print("Considering constant substitutions: " + key + " with " + value)
	requestbulk = requestbulk.replace(key, value)
	if args.verbose:
		print ("Replacement in constant request: replaced {0} with {1}".format(key, value))

#parse the bulk and get a list of requests
request_list = re.split(separator, requestbulk)

#payloads parsing if any	
if 'payloads' in config.keys():
	iterations = parse_payloads()
else:
	iterations = 1

#initializing response vars and assigning them to the appropriate steps
step_substitutions = [dict() for x in range(len(request_list))]
current_step_substitutions = {}
step_substitutions_num = 0
if ('substitutions' in config.keys()) and ('responsevars' in config['substitutions'].keys()):
    for key,value in config['substitutions']['responsevars'].items():
        if 'json' in value.keys():
            if 'steps' in value.keys():
                for i in value['steps']:
                    step_substitutions[i][key] = ('json',value['json'])
            else:
                for i in range(0,len(request_list),1):
                    step_substitutions[i][key] = ('json',value['json'])
        elif 'regex' in value.keys():
            try:
                re.compile(value['regex'])
            except Exception as e:
                print ("Can't compile regex {0}, skipping and, continuing.".format(value['regex']))
                continue
            if 'steps' in value.keys():
                for i in value['steps']:
                    step_substitutions[i][key] = ('regex',value['regex'])
            else:
                for i in range(0,len(request_list),1):
                    step_substitutions[i][key] = ('regex',value['regex'])
        else:
            print("Error - can't parse config file - invalid responsevar, exiting.")
            if logfile:
                logfile.close()
            sys.exit(1)
        step_substitutions_num += 1

if 'grep_last_response' in config and config['grep_last_response']:
	try:
		re.compile(config['grep_last_response'])
		grep_last_response_regex = config['grep_last_response']
	except Exception as e:
		print ("Error - can't compile grep_last_response regex {0}, exiting.".format(value['regex']))
		sys.exit(1)

print ("Parsed: {0} constant substitutions, {1} response variables and {2} injection points; payload sequence length is {3}; total requests to be sent {4}.".format(
	len(constant_substitutions),
	step_substitutions_num,
	len(payloads), 
	iterations,
	len(request_list)*iterations))

#main sending loop
for i in range(0, iterations, 1):
	send_sequence(request_list, i)
	
if logfile:
	logfile.close()